src/Components/Blazor/Validation/src/ObjectGraphDataAnnotationsValidator.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    public class ObjectGraphDataAnnotationsValidator : ComponentBase
+    {
+        private static readonly object ValidationContextValidatorKey = new object();
+        private static readonly object ValidatedObjectsKey = new object();
+        private ValidationMessageStore _validationMessageStore;
+
+        [CascadingParameter]
+        internal EditContext EditContext { get; set; }
+
+        protected override void OnInitialized()
+        {
+            _validationMessageStore = new ValidationMessageStore(EditContext);
+
+            // Perform object-level validation (starting from the root model) on request
+            EditContext.OnValidationRequested += (sender, eventArgs) =>
+            {
+                _validationMessageStore.Clear();
+                ValidateObject(EditContext.Model, new HashSet<object>());
+                EditContext.NotifyValidationStateChanged();
+            };
+
+            // Perform per-field validation on each field edit
+            EditContext.OnFieldChanged += (sender, eventArgs) =>
+                ValidateField(EditContext, _validationMessageStore, eventArgs.FieldIdentifier);
+        }
+
+        internal void ValidateObject(object value, HashSet<object> visited)
+        {
```

Comment:
Do we have a test for this that validates object graphs with cycles? (I haven't been able to find it).

Reply:
I added a couple here: https://github.com/aspnet/AspNetCore/pull/14972/files#diff-a3fb926db31109f966f332738bc6f29bR309-R426